import{_ as s}from"./plugin-vue_export-helper-DlAUqK2U.js";import{c as e,b as a,o as n}from"./app-ptbqs2O5.js";const t={};function l(p,i){return n(),e("div",null,i[0]||(i[0]=[a(`<p>为了避免程序在出现异常时直接退出，我们可以使用<code>try-except</code>语句捕获可能发生的异常。</p><h2 id="基本结构" tabindex="-1"><a class="header-anchor" href="#基本结构"><span>基本结构</span></a></h2><p><code>try-except</code>的基本结构就像这样：</p><div class="language-python line-numbers-mode" data-highlighter="shiki" data-ext="python" data-title="python" style="--shiki-light:#383A42;--shiki-dark:#abb2bf;--shiki-light-bg:#FAFAFA;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes one-light one-dark-pro vp-code"><code><span class="line"><span style="--shiki-light:#A626A4;--shiki-dark:#C678DD;">try</span><span style="--shiki-light:#383A42;--shiki-dark:#ABB2BF;">:</span></span>
<span class="line"><span style="--shiki-light:#0184BC;--shiki-dark:#56B6C2;">    print</span><span style="--shiki-light:#383A42;--shiki-dark:#ABB2BF;">(</span><span style="--shiki-light:#50A14F;--shiki-dark:#98C379;">&quot;这是可能发生错误的代码&quot;</span><span style="--shiki-light:#383A42;--shiki-dark:#ABB2BF;">)</span></span>
<span class="line"><span style="--shiki-light:#383A42;--shiki-dark:#ABB2BF;">    a </span><span style="--shiki-light:#383A42;--shiki-dark:#56B6C2;">=</span><span style="--shiki-light:#986801;--shiki-dark:#D19A66;"> 123</span><span style="--shiki-light:#383A42;--shiki-dark:#56B6C2;">/</span><span style="--shiki-light:#986801;--shiki-dark:#D19A66;">0</span></span>
<span class="line"><span style="--shiki-light:#A626A4;--shiki-dark:#C678DD;">except</span><span style="--shiki-light:#383A42;--shiki-dark:#ABB2BF;">:</span></span>
<span class="line"><span style="--shiki-light:#0184BC;--shiki-dark:#56B6C2;">    print</span><span style="--shiki-light:#383A42;--shiki-dark:#ABB2BF;">(</span><span style="--shiki-light:#50A14F;--shiki-dark:#98C379;">&quot;发生错误&quot;</span><span style="--shiki-light:#383A42;--shiki-dark:#ABB2BF;">)</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>这段代码试图将某个值除以0，但是在运行时没有抛出异常，而是执行了except下方的代码，打印了<code>发生错误</code>。</p><p>注意：在执行过<code>except</code>下方语句后，不会再继续执行<code>try</code>下方的语句。</p><h2 id="捕获特定异常" tabindex="-1"><a class="header-anchor" href="#捕获特定异常"><span>捕获特定异常</span></a></h2><p>您可以让<code>except</code>子句仅捕获特定的异常，就像这样：</p><div class="language-python line-numbers-mode" data-highlighter="shiki" data-ext="python" data-title="python" style="--shiki-light:#383A42;--shiki-dark:#abb2bf;--shiki-light-bg:#FAFAFA;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes one-light one-dark-pro vp-code"><code><span class="line"><span style="--shiki-light:#A626A4;--shiki-dark:#C678DD;">try</span><span style="--shiki-light:#383A42;--shiki-dark:#ABB2BF;">:</span></span>
<span class="line"><span style="--shiki-light:#0184BC;--shiki-dark:#56B6C2;">    print</span><span style="--shiki-light:#383A42;--shiki-dark:#ABB2BF;">(</span><span style="--shiki-light:#50A14F;--shiki-dark:#98C379;">&quot;这是可能发生错误的代码&quot;</span><span style="--shiki-light:#383A42;--shiki-dark:#ABB2BF;">)</span></span>
<span class="line"><span style="--shiki-light:#383A42;--shiki-dark:#ABB2BF;">    a </span><span style="--shiki-light:#383A42;--shiki-dark:#56B6C2;">=</span><span style="--shiki-light:#986801;--shiki-dark:#D19A66;"> 123</span><span style="--shiki-light:#383A42;--shiki-dark:#56B6C2;"> &lt;</span><span style="--shiki-light:#50A14F;--shiki-dark:#98C379;"> &quot;string&quot;</span></span>
<span class="line"><span style="--shiki-light:#A626A4;--shiki-dark:#C678DD;">except</span><span style="--shiki-light:#0184BC;--shiki-dark:#ABB2BF;"> ZeroDivisionError</span><span style="--shiki-light:#383A42;--shiki-dark:#ABB2BF;">:</span></span>
<span class="line"><span style="--shiki-light:#0184BC;--shiki-dark:#56B6C2;">    print</span><span style="--shiki-light:#383A42;--shiki-dark:#ABB2BF;">(</span><span style="--shiki-light:#50A14F;--shiki-dark:#98C379;">&quot;不允许除以0&quot;</span><span style="--shiki-light:#383A42;--shiki-dark:#ABB2BF;">)</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>这段代码会在发生<code>ZeroDivisionError</code>时输出<code>不允许除以0</code>，但是在遇到其它错误时仍然会抛出异常，就像这样：</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#383A42;--shiki-dark:#abb2bf;--shiki-light-bg:#FAFAFA;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes one-light one-dark-pro vp-code"><code><span class="line"><span>Traceback (most recent call last):</span></span>
<span class="line"><span>  File &quot;d:\\test.py&quot;, line 4, in &lt;module&gt;</span></span>
<span class="line"><span>    a = 123 &lt; &quot;string&quot;</span></span>
<span class="line"><span>        ^^^^^^^^^^^^^^</span></span>
<span class="line"><span>TypeError: &#39;&lt;&#39; not supported between instances of &#39;int&#39; and &#39;str&#39;</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>此时虽然发生了错误，但是由于我们没有要求<code>try-except</code>语句对这种异常进行捕获，因此抛出了异常。</p><h2 id="获取异常信息" tabindex="-1"><a class="header-anchor" href="#获取异常信息"><span>获取异常信息</span></a></h2><p>您可以通过在<code>except</code>语句中使用<code>as</code>来获取异常信息，就像这样：</p><div class="language-python line-numbers-mode" data-highlighter="shiki" data-ext="python" data-title="python" style="--shiki-light:#383A42;--shiki-dark:#abb2bf;--shiki-light-bg:#FAFAFA;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes one-light one-dark-pro vp-code"><code><span class="line"><span style="--shiki-light:#A626A4;--shiki-dark:#C678DD;">try</span><span style="--shiki-light:#383A42;--shiki-dark:#ABB2BF;">:</span></span>
<span class="line"><span style="--shiki-light:#0184BC;--shiki-dark:#56B6C2;">    print</span><span style="--shiki-light:#383A42;--shiki-dark:#ABB2BF;">(</span><span style="--shiki-light:#50A14F;--shiki-dark:#98C379;">&quot;这是可能发生错误的代码&quot;</span><span style="--shiki-light:#383A42;--shiki-dark:#ABB2BF;">)</span></span>
<span class="line"><span style="--shiki-light:#383A42;--shiki-dark:#ABB2BF;">    a </span><span style="--shiki-light:#383A42;--shiki-dark:#56B6C2;">=</span><span style="--shiki-light:#986801;--shiki-dark:#D19A66;"> 123</span><span style="--shiki-light:#383A42;--shiki-dark:#56B6C2;"> &lt;</span><span style="--shiki-light:#50A14F;--shiki-dark:#98C379;"> &quot;string&quot;</span></span>
<span class="line"><span style="--shiki-light:#A626A4;--shiki-dark:#C678DD;">except</span><span style="--shiki-light:#0184BC;--shiki-dark:#ABB2BF;"> TypeError</span><span style="--shiki-light:#A626A4;--shiki-dark:#C678DD;"> as</span><span style="--shiki-light:#383A42;--shiki-dark:#ABB2BF;"> e:</span></span>
<span class="line"><span style="--shiki-light:#0184BC;--shiki-dark:#56B6C2;">    print</span><span style="--shiki-light:#383A42;--shiki-dark:#ABB2BF;">(e)</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div><p>其执行结果如下：</p><div class="language- line-numbers-mode" data-highlighter="shiki" data-ext="" data-title="" style="--shiki-light:#383A42;--shiki-dark:#abb2bf;--shiki-light-bg:#FAFAFA;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes one-light one-dark-pro vp-code"><code><span class="line"><span>这是可能发生错误的代码</span></span>
<span class="line"><span>&#39;&lt;&#39; not supported between instances of &#39;int&#39; and &#39;str&#39;</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div></div></div><p>此时虽然打印了报错信息，但是程序并没有因此终止运行。</p><p>注意：如果您需要使用<code>as</code>来获取报错信息，则必须指定需要捕获的异常。</p><p>如果您不知道这段代码会发生什么问题，请指定捕获<code>Exception</code>，这相当于捕获所有异常，就像这样：</p><div class="language-python line-numbers-mode" data-highlighter="shiki" data-ext="python" data-title="python" style="--shiki-light:#383A42;--shiki-dark:#abb2bf;--shiki-light-bg:#FAFAFA;--shiki-dark-bg:#282c34;"><pre class="shiki shiki-themes one-light one-dark-pro vp-code"><code><span class="line"><span style="--shiki-light:#A626A4;--shiki-dark:#C678DD;">try</span><span style="--shiki-light:#383A42;--shiki-dark:#ABB2BF;">:</span></span>
<span class="line"><span style="--shiki-light:#0184BC;--shiki-dark:#56B6C2;">    print</span><span style="--shiki-light:#383A42;--shiki-dark:#ABB2BF;">(</span><span style="--shiki-light:#50A14F;--shiki-dark:#98C379;">&quot;这是可能发生错误的代码&quot;</span><span style="--shiki-light:#383A42;--shiki-dark:#ABB2BF;">)</span></span>
<span class="line"><span style="--shiki-light:#383A42;--shiki-dark:#ABB2BF;">    a </span><span style="--shiki-light:#383A42;--shiki-dark:#56B6C2;">=</span><span style="--shiki-light:#986801;--shiki-dark:#D19A66;"> 123</span><span style="--shiki-light:#383A42;--shiki-dark:#56B6C2;"> &lt;</span><span style="--shiki-light:#50A14F;--shiki-dark:#98C379;"> &quot;string&quot;</span></span>
<span class="line"><span style="--shiki-light:#A626A4;--shiki-dark:#C678DD;">except</span><span style="--shiki-light:#0184BC;--shiki-dark:#ABB2BF;"> Exception</span><span style="--shiki-light:#A626A4;--shiki-dark:#C678DD;"> as</span><span style="--shiki-light:#383A42;--shiki-dark:#ABB2BF;"> e:</span></span>
<span class="line"><span style="--shiki-light:#0184BC;--shiki-dark:#56B6C2;">    print</span><span style="--shiki-light:#383A42;--shiki-dark:#ABB2BF;">(e)</span></span></code></pre><div class="line-numbers" aria-hidden="true" style="counter-reset:line-number 0;"><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div><div class="line-number"></div></div></div>`,21)]))}const d=s(t,[["render",l],["__file","try-except-basic.html.vue"]]),k=JSON.parse('{"path":"/docs/python-exception-handling/try-except-basic.html","title":"try-except 语句基础","lang":"zh-CN","frontmatter":{"title":"try-except 语句基础","tag":"Python, 异常处理：让程序更稳健, try-except","author":"灵魂歌手er","description":"为了避免程序在出现异常时直接退出，我们可以使用try-except语句捕获可能发生的异常。 基本结构 try-except的基本结构就像这样： 这段代码试图将某个值除以0，但是在运行时没有抛出异常，而是执行了except下方的代码，打印了发生错误。 注意：在执行过except下方语句后，不会再继续执行try下方的语句。 捕获特定异常 您可以让excep...","head":[["meta",{"property":"og:url","content":"https://py.hello8693.xyz/docs/python-exception-handling/try-except-basic.html"}],["meta",{"property":"og:site_name","content":"Python 教程"}],["meta",{"property":"og:title","content":"try-except 语句基础"}],["meta",{"property":"og:description","content":"为了避免程序在出现异常时直接退出，我们可以使用try-except语句捕获可能发生的异常。 基本结构 try-except的基本结构就像这样： 这段代码试图将某个值除以0，但是在运行时没有抛出异常，而是执行了except下方的代码，打印了发生错误。 注意：在执行过except下方语句后，不会再继续执行try下方的语句。 捕获特定异常 您可以让excep..."}],["meta",{"property":"og:type","content":"article"}],["meta",{"property":"og:locale","content":"zh-CN"}],["meta",{"property":"og:updated_time","content":"2024-12-29T00:40:29.000Z"}],["meta",{"property":"article:author","content":"灵魂歌手er"}],["meta",{"property":"article:tag","content":"Python, 异常处理：让程序更稳健, try-except"}],["meta",{"property":"article:modified_time","content":"2024-12-29T00:40:29.000Z"}],["script",{"type":"application/ld+json"},"{\\"@context\\":\\"https://schema.org\\",\\"@type\\":\\"Article\\",\\"headline\\":\\"try-except 语句基础\\",\\"image\\":[\\"\\"],\\"dateModified\\":\\"2024-12-29T00:40:29.000Z\\",\\"author\\":[{\\"@type\\":\\"Person\\",\\"name\\":\\"灵魂歌手er\\"}]}"]]},"headers":[{"level":2,"title":"基本结构","slug":"基本结构","link":"#基本结构","children":[]},{"level":2,"title":"捕获特定异常","slug":"捕获特定异常","link":"#捕获特定异常","children":[]},{"level":2,"title":"获取异常信息","slug":"获取异常信息","link":"#获取异常信息","children":[]}],"git":{"createdTime":1734756596000,"updatedTime":1735432829000,"contributors":[{"name":"LHGS-github","username":"LHGS-github","email":"340994706@qq.com","commits":1,"url":"https://github.com/LHGS-github"},{"name":"unDefFtr","username":"unDefFtr","email":"undefftr@undefined.ac.cn","commits":1,"url":"https://github.com/unDefFtr"},{"name":"hello8693","username":"hello8693","email":"1320998105@qq.com","commits":1,"url":"https://github.com/hello8693"}]},"readingTime":{"minutes":1.71,"words":512},"filePathRelative":"docs/python-exception-handling/try-except-basic.md","localizedDate":"2024年12月21日","autoDesc":true}');export{d as comp,k as data};
